File: Assets/Animations/Animation.ts
import { isEmpty } from "Helpers";
import Asset from "../Asset";
import FontWithPosition from "Assets/Fonts/FontWithPosition";
import { Source } from "love.audio";
import GameManager from "GameManager";

export interface AnimationOptions {
  readonly animDuration?: number;
  readonly onFinish?: () => void;
  readonly waitForAnimationIds?: string[];
  readonly stopAnimationCondition?: () => boolean;
  readonly soundToPlay?: Source;
}

export type AnimationAssets = Asset | FontWithPosition;

export default abstract class Animation {
  gameManager: GameManager;
  id: string;
  animDuration?: number;
  animElapsed = 0;
  isAnimating = false;
  hasStarted = false;
  assets: AnimationAssets[];
  originalX: Map<string, number> = new Map<string, number>();
  originalY: Map<string, number> = new Map<string, number>();
  onFinish?: () => void;
  waitForAnimationIds: string[];
  stopAnimationCondition?: () => boolean;
  soundToPlay?: Source;
  playedSound = false;
  isPaused = false;

  constructor(
    gameManager: GameManager,
    id: string,
    assets: AnimationAssets[],
    constructionOptions?: AnimationOptions
  ) {
    this.gameManager = gameManager;
    this.id = id;
    this.animDuration = constructionOptions?.animDuration;
    this.animElapsed = 0;
    this.isAnimating = true;
    this.assets = assets;
    this.originalX = new Map(assets.map((asset) => [asset.id, asset.x]));
    this.originalY = new Map(assets.map((asset) => [asset.id, asset.y]));
    this.onFinish = constructionOptions?.onFinish;
    this.waitForAnimationIds = constructionOptions?.waitForAnimationIds ?? [];
    this.stopAnimationCondition = constructionOptions?.stopAnimationCondition;
    this.soundToPlay = constructionOptions?.soundToPlay;
  }

  updateAnimation(deltaTime: number): void {
    this.hasStarted = true;

    if (!this.isAnimating) {
      return;
    }

    if (this.stopAnimationCondition && this.stopAnimationCondition()) {
      this.isAnimating = false;
      return;
    }

    this.animElapsed += deltaTime;

    if (!isEmpty(this.animDuration) && this.animElapsed >= this.animDuration) {
      // Animation complete
      this.animElapsed = this.animDuration;
      this.isAnimating = false;
    }
  }

  get isFinished(): boolean {
    return !this.isAnimating;
  }

  getAssets(): AnimationAssets[] {
    return this.assets;
  }

  updateX(deltaX: number): void {
    this.assets.forEach((asset) => {
      const originalX = this.originalX.get(asset.id);
      if (originalX !== undefined) {
        asset.x = originalX + deltaX;
      }
    });
  }

  updateY(deltaY: number): void {
    this.assets.forEach((asset) => {
      const originalY = this.originalY.get(asset.id);
      if (originalY !== undefined) {
        asset.y = originalY + deltaY;
      }
    });
  }

  shouldWaitForAnimations(): boolean {
    if (this.waitForAnimationIds.length > 0) {
      for (const waitId of this.waitForAnimationIds) {
        if (waitId === this.id) {
          continue;
        }

        const waitForAnimation =
          this.gameManager.animationManager.animations.get(waitId);
        if (
          !isEmpty(waitForAnimation) &&
          !waitForAnimation.isFinished &&
          waitForAnimation.hasStarted
        ) {
          return true;
        }
      }
    }
    return false;
  }
}

File: Assets/Animations/Animation.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ArrayMap = ____lualib.__TS__ArrayMap
local Map = ____lualib.Map
local __TS__New = ____lualib.__TS__New
local __TS__ArrayForEach = ____lualib.__TS__ArrayForEach
local __TS__SetDescriptor = ____lualib.__TS__SetDescriptor
local ____exports = {}
local ____Helpers = require("Helpers")
local isEmpty = ____Helpers.isEmpty
____exports.default = __TS__Class()
local Animation = ____exports.default
Animation.name = "Animation"
function Animation.prototype.____constructor(self, gameManager, id, assets, constructionOptions)
    self.animElapsed = 0
    self.isAnimating = false
    self.hasStarted = false
    self.originalX = __TS__New(Map)
    self.originalY = __TS__New(Map)
    self.playedSound = false
    self.isPaused = false
    self.gameManager = gameManager
    self.id = id
    self.animDuration = constructionOptions and constructionOptions.animDuration
    self.animElapsed = 0
    self.isAnimating = true
    self.assets = assets
    self.originalX = __TS__New(
        Map,
        __TS__ArrayMap(
            assets,
            function(____, asset) return {asset.id, asset.x} end
        )
    )
    self.originalY = __TS__New(
        Map,
        __TS__ArrayMap(
            assets,
            function(____, asset) return {asset.id, asset.y} end
        )
    )
    self.onFinish = constructionOptions and constructionOptions.onFinish
    self.waitForAnimationIds = constructionOptions and constructionOptions.waitForAnimationIds or ({})
    self.stopAnimationCondition = constructionOptions and constructionOptions.stopAnimationCondition
    self.soundToPlay = constructionOptions and constructionOptions.soundToPlay
end
function Animation.prototype.updateAnimation(self, deltaTime)
    self.hasStarted = true
    if not self.isAnimating then
        return
    end
    if self.stopAnimationCondition and self:stopAnimationCondition() then
        self.isAnimating = false
        return
    end
    self.animElapsed = self.animElapsed + deltaTime
    if not isEmpty(self.animDuration) and self.animElapsed >= self.animDuration then
        self.animElapsed = self.animDuration
        self.isAnimating = false
    end
end
function Animation.prototype.getAssets(self)
    return self.assets
end
function Animation.prototype.updateX(self, deltaX)
    __TS__ArrayForEach(
        self.assets,
        function(____, asset)
            local originalX = self.originalX:get(asset.id)
            if originalX ~= nil then
                asset.x = originalX + deltaX
            end
        end
    )
end
function Animation.prototype.updateY(self, deltaY)
    __TS__ArrayForEach(
        self.assets,
        function(____, asset)
            local originalY = self.originalY:get(asset.id)
            if originalY ~= nil then
                asset.y = originalY + deltaY
            end
        end
    )
end
function Animation.prototype.shouldWaitForAnimations(self)
    if #self.waitForAnimationIds > 0 then
        for ____, waitId in ipairs(self.waitForAnimationIds) do
            do
                if waitId == self.id then
                    goto __continue18
                end
                local waitForAnimation = self.gameManager.animationManager.animations:get(waitId)
                if not isEmpty(waitForAnimation) and not waitForAnimation.isFinished and waitForAnimation.hasStarted then
                    return true
                end
            end
            ::__continue18::
        end
    end
    return false
end
__TS__SetDescriptor(
    Animation.prototype,
    "isFinished",
    {get = function(self)
        return not self.isAnimating
    end},
    true
)
return ____exports
